drivers/regulator/regulator_gpio.c

```diff
@@ -0,0 +1,187 @@
+		}
+
+		diff = mid - cfg->states[state * 2];
+		if (diff < 0) {
+			diff = -diff;
+		}
+
+		if (diff < min_diff) {
+			min_diff = diff;
+			best_state = state;
+		}
+	}
+
+	if (min_diff == INT32_MAX) {
+		LOG_ERR("%s: can't find voltage is states", dev->name);
+		return -EINVAL;
+	}
+
+	ret = gpio_pin_get_dt(&cfg->switch_volt);
+	if (ret < 0) {
+		LOG_ERR("%s: can't get pin state", dev->name);
+		return ret;
+	}
+
+	if (ret != cfg->states[best_state * 2 + 1]) {
+		ret = gpio_pin_set_dt(&cfg->switch_volt, cfg->states[best_state * 2 + 1]);
+		if (ret < 0) {
+			LOG_ERR("%s: can't set pin state", dev->name);
+			return ret;
+		}
+	}
+
+	data->current_volt_uv = cfg->states[best_state * 2];
+	return 0;
+}
+
+static int regulator_gpio_get_voltage(const struct device *dev, int32_t *volt_uv)
+{
+	const struct regulator_gpio_data *data = dev->data;
+
+	*volt_uv = data->current_volt_uv;
+	return 0;
+}
+
+static const struct regulator_driver_api regulator_gpio_api = {
+	.set_voltage = regulator_gpio_set_voltage,
+	.get_voltage = regulator_gpio_get_voltage,
+	.count_voltages = regulator_gpio_count_voltages,
+	.list_voltage = regulator_gpio_list_voltage,
+};
+
+static int regulator_gpio_init(const struct device *dev)
+{
+	const struct regulator_gpio_config *cfg = dev->config;
+	struct regulator_gpio_data *data = dev->data;
+	int ret;
+
+	regulator_common_data_init(dev);
+
+	if (!device_is_ready(cfg->switch_volt.port)) {
+		LOG_ERR("%s: gpio port: %s not ready", dev->name, cfg->switch_volt.port->name);
+		return -ENODEV;
+	}
+
+	ret = gpio_pin_configure_dt(&cfg->switch_volt,
+				    cfg->initial_state ? GPIO_OUTPUT_ACTIVE :
+							 GPIO_OUTPUT_INACTIVE);
+	if (ret < 0) {
+		LOG_ERR("%s: can't configure pin", dev->name);
+		return ret;
+	}
+
+	ret = regulator_gpio_get_state_voltage(dev, cfg->initial_state, &data->current_volt_uv);
+	if (ret < 0) {
+		LOG_ERR("%s: can't get list volatge for state %d", dev->name, cfg->initial_state);
+		return ret;
+	}
+
+	ret = regulator_common_init(dev, true);
+	if (ret < 0) {
+		LOG_ERR("%s: can't init common part of regulator", dev->name);
+		return ret;
+	}
+
+	return 0;
+}
+
+#define REGULATOR_GPIO_DEFINE(inst) \
+	static struct regulator_gpio_data data##inst; \
+	BUILD_ASSERT(!(DT_PROP_LEN(DT_DRV_INST(inst), states) & 0x1), \
+		     "Number of regulator states should be even"); \
+	static const struct regulator_gpio_config config##inst = { \
+		.common = REGULATOR_DT_INST_COMMON_CONFIG_INIT(inst), \
+		.switch_volt = GPIO_DT_SPEC_INST_GET(inst, switch_voltage_gpios), \
+		.initial_state = DT_PROP(DT_DRV_INST(inst), gpios_state), \
+		.states = ((const int []) \
+			      DT_PROP(DT_DRV_INST(inst), states)), \
+		.states_cnt = DT_PROP_LEN(DT_DRV_INST(inst), states), \
+	}; \
+	DEVICE_DT_INST_DEFINE(inst, regulator_gpio_init, NULL, &data##inst, \
+			      &config##inst, POST_KERNEL, \
+			      CONFIG_REGULATOR_GPIO_INIT_PRIORITY, \
+			      &regulator_gpio_api);
+
+DT_INST_FOREACH_STATUS_OKAY(REGULATOR_GPIO_DEFINE)
```

Comment:
please format this properly, try with ` clang-format`

Reply:
could you please explain what do mean under `properly`?

Reply:
https://docs.zephyrproject.org/latest/contribute/guidelines.html#clang-format